### ISSUE TITLE
Gizmo is unusable

### ISSUE THREAD

**Issue by HaSa1002** [Libre-TrainSim/Libre-TrainSim#166]
**See https://github.com/Libre-TrainSim/Libre-TrainSim/issues/166#issuecomment-1248720253 for an updated concept**

The editor gizmo is unusable. Here is a list of issues:
- [x] #415
- [x] Roationgizmo should be a sphere (or circle)
- [x] Blink effect should not affect the object itself (only the gizmo) (or don't blink at all)
- [x] Movement gizmo should be the same size independent of the distance to the camera
- [x] Gizmos should have a safety selection margin
- [x] #416
- [ ] Objects should be moved on xz Plane and y should be automatically determined by terrain or other objects

**Comment by DasCapschen** [comment]
size independence is a material parameter, if I recall correctly. So fixing that should literally be one click :smile:

**Comment by HaSa1002** [comment]
Yeah, but that doesn't have the effect I was looking for. At least for now. Let's see whoever gets to fiddle with this issue ^^

**Comment by HaSa1002** [comment]
Idea: We could draw the stuff directly on the screen using a determined viewport and sync the rotation around the gizmo with the camera. Non-optimal, but I am clueless how to this properly atm. Probably involves using a ImmediateGeometry node

**Comment by HaSa1002** [comment]
It can be entirely done using 3D nodes. I made the following sketch showcasing a more usable gizmo. This concept remains close to a traditional gizmo. I think object picking should be implemented together with "auto-height" and left out in the inital rework.

![gizmo](https://user-images.githubusercontent.com/14185889/190476156-c691f499-30ea-4aa1-b84b-9b8bcdbb55a1.png)

## Description

The blue arrows (direction arrows) are the arrows for each direction. They have a grey click margin. The purple double arrows (rotation arrows) are used to rotate. They too have a click margin. When clicking into the red area, all movement is on the the selected plane. (Two axis at the same time). The red plane is not visible by default.

## Visual clues

The direction arrows are coloured in their *axis colour*. **These still need to be determined but should not include green because of our terrain and red-green color blindness being very common.**
The rotation arrows are coloured in the *axis colour of the axis they're rotating around*.
When hovering above an arrow, its colour should saturate. (or lighten?)
Additonally the other arrows should be desaturated (or darkend), depending on the distance to the cursor.
Above a certain threshold, the plane should be shown along with the saturation of both arrows.
There needs to be some sort of indication of the coordinate space. **TBD**
The current position and rotation should be visible next to their axis.
When the axis is changed (moving or rotation) all other values should dissapear and the important values should also show the difference to the start of the operation.
When an operation starts, all irrelevant indicators dissapear.
The difference to the start values should be visually visible as well. The rotation likely needs some sort of start indicator. The movement probably needs some sort of "cross" staying at the start position.

## Additional functionality

Pressing control while moving snaps to increments of 0.25, 1, 10, or 25 in global coordinates depending on the zoom level or closeness to the object at the beginning of the operation.
Pressing control while rotating snaps the rotation to increments of 5, 10, 45, or 90 degrees in global space depending on the distance to the object or zoom level at the beginning of the operation.
When pressing shift, the mouse movement as well as snapping is lowered by factor ten.
Pressing alt switches the operation to local space.
Pressing escape or the right mouse button cancels the operation.

## Object picking

Object picking complements a good gizmo. Picked objects stick to the cursor. A gizmo is better suited for fine tuning while object picking is more useful when roughly setting the position of objects. This boils down to defaulting to object picking when starting placement of new objects. Additionally it can be activated by pressing a button (p?) and perhaps a button on the gizmo. While moving a picked object, it should snap to other objects and rails. This behaviour can be turned off by pressing control. Pressing shift temporarily slows the relative movement of the mouse down, meaning the object moves slower than the mouse and builds up a distance to it. Thus, on release of shift, the object needs to be reset to the cursor position. It enables a little more fine grained control before placing the object down by pressing the left mouse button. Pressing escape (not the right mouse button, because of the FP camera) cancels the operation.

**Object picking needs more design work and thought.**

## Closing thoughts

This is just a design concept. Some parts are brain verifiable but most probably some testing. I am not done thinking about object picking and it will be split into its own proposal but must be thought in tandem with a good gizmo. I've some experience with gizmos but every single one has different use cases and needs different tools to enable different workflows. I wrote it down to enable discussion of it and implementation by someone not being myself. Please share your thoughts.
